### DISCUSSION TITLE
copy to hangs for no apparent reason on large transfers ~80% of the way through

### DISCUSSION THREAD

**Discussion by LukeMckim** [psycopg/psycopg#587] [Q&A]
I've had the occur a few times with different tables where when I'm trying to transfer alot of data from one server to another the operation hangs on the `write_row` portion about 80% of the way through.

For a table with >50M rows this would occur ~42M, and recently on a table with >9M rows it occurs somewhere between 8.1M and 8.7M rows.
Whats odd to me is that it doesnt occur at the same point everytime but does happen at _some_ point everytime. 
I've checked my timeout parameters and their set beyond whats needed.
I've also checked the databases pg_stat_activity and the source says its waiting on ClientRead, while the destination is stuck in DataFileRead with wait_type = IO, but this never changes.
I've also tried not using `write_rows` and instead used just `read` and `write` but no luck there either.
I'm pretty stuck at this point, any ideas?

my code is:
```python
async with src_conn.cursor(f"{table}_full_load_out") as src_acsr:
    async with dst_conn.cursor(f"{table}_full_load_in") as dst_acsr:
        async with src_acsr.copy(f"COPY (SELECT * FROM {table}) TO STDOUT") as cpy_from:
            async with dst_acsr.copy(f"COPY {table} FROM STDIN") as cpy_to:
                async for row in cpy_from.rows():
                    await cpy_to.write_row(row)
```

**Comment by LukeMckim** [comment]
I'm going to add some findings I made to continue progressing in my project in case anyone else stumbles upon this.

I couldn't figure out why it was hanging, both the source and destination servers said the connections were still active but nothing appeared to happen no matter how long I left it.

Instead I found 2 things that finally allowed me to complete a full transfer on these larger tables.
The first was to break the copy to portion into blocks instead of trying to do it all at once. By storing the `read` results into a batch (100K) and then opening the copy to cursor and copying those results and clearing the batch I was able to break the past the limits I was hitting before. However it was still very slow, much slower than the initial attempt.

Next I set it up so that I dropped the constraints/indexes on the destination table before the load. This sped things up greatly. I would then just rebuild the indexes after the transfer was complete.

Hopefully this is helpful for someone else.

**Reply by dvarrazzo** [reply]
Hello,

sorry, I couldn't answer in the previous days, but I wanted to get back on this.

I didn't see, in your code, any problem. So I assume the problem is in the data. I wanted to ask you if you could provide a sample data for testing, even anonymised of course: if any there might be an issue caused by the data size (row size? field size? That's would I would have investigated).

**Reply by LukeMckim** [reply]
I cant provide the actual data itself but I can provide the two table schemas where I've encountered the errors. Just to clarify as well I was checking the data rows that the copy instruction was getting stuck on and nothing appeared out of the ordinary, no longer than average fields or unexpected data types. Also it was getting stuck in different places and other copy operations would work such as my solution laid out above and also doing a direct `psql -c ` COPY piping from one server to the other.

this is the one that was breaking down ~8M/9M rows
```
          Column           |           Type           | Collation | Nullable | Default 
---------------------------+--------------------------+-----------+----------+---------
 col_a                     | uuid                     |           | not null | 
 col_b                     | timestamp with time zone |           |          | 
 col_c                     | uuid                     |           |          | 
 col_d                     | text                     |           |          | 
 col_e                     | side                     |           |          | 
 col_f                     | text                     |           |          | 
 col_g                     | text                     |           |          | 
 col_h                     | numeric                  |           |          | 
 col_i                     | numeric                  |           |          | 
 col_j                     | numeric                  |           |          | 
 col_k                     | numeric                  |           |          | 
 col_l                     | quote_status             |           |          | 
 col_m                     | timestamp with time zone |           |          | 
 col_n                     | jsonb                    |           |          | 
 col_o                     | jsonb                    |           |          | 
 col_p                     | text                     |           |          | 
Indexes:
    "table_pkey" PRIMARY KEY, btree (uuid)
    "idx_table_col_c_btree" btree (col_c)
```

and this is the one that would breakdown ~40M/50M rows
```
        Column        |            Type             | Collation | Nullable | Default 
----------------------+-----------------------------+-----------+----------+---------
 col_a                | uuid                        |           | not null | 
 col_b                | uuid                        |           |          | 
 col_c                | uuid                        |           |          | 
 col_d                | uuid                        |           |          | 
 col_e                | character varying(255)      |           |          | 
 col_f                | uuid                        |           |          | 
 col_g                | character varying(100)      |           |          | 
 col_h                | character varying(100)      |           |          | 
 col_i                | character varying(100)      |           |          | 
 col_j                | character varying(100)      |           |          | 
 col_k                | character varying(100)      |           |          | 
 col_l                | character varying(100)      |           |          | 
 col_m                | character varying(100)      |           |          | 
 col_n                | character varying(36)       |           |          | 
 col_o                | boolean                     |           |          | 
 col_p                | character varying(100)      |           |          | 
 col_q                | character varying(100)      |           |          | 
 col_r                | timestamp without time zone |           |          | 
 col_s                | uuid                        |           |          | 
 col_t                | character varying(100)      |           |          | 
 col_u                | numeric(7,4)                |           |          | 
 col_a                | numeric(7,4)                |           |          | 
 col_v                | character varying           |           |          | 
 col_w                | text                        |           |          | 
 col_x                | boolean                     |           |          | 
 col_y                | boolean                     |           |          | 
 col_z                | boolean                     |           |          | 
 col_0                | text                        |           |          | 
 col_1                | text                        |           |          | 
 col_2                | text                        |           |          | 
 col_3                | text                        |           |          | 
 col_4                | text                        |           |          | 
 col_5                | text                        |           |          | 
 col_6                | boolean                     |           |          | 
 col_7                | text                        |           |          | 
 col_8                | boolean                     |           |          | 
 col_9                | timestamp without time zone |           |          | 
Indexes:
    "table_col_a_idx" btree (col_a)
```

these are the schemas of the destination tables that were being loaded **into**.
I appreciate any and all help and thanks for all your hard work
